// @omnistudio/resources/src/javascript/utils/webpackBuilder.js
const { readdirSync, statSync, existsSync } = require('fs'),
    path = require('path');

function getFilesFromPath(basePath, relativePath) {
    let results = [];

    if (existsSync(basePath)) {

        // Get the list of files/directories from the path
        const list = readdirSync(basePath);

        // Loop into each one to get the files and recursive find into folders
        list.forEach(p => {
            const realPath = basePath + '/' + p,
                info = statSync(realPath);
            if (info.isDirectory()) {

                const child = getFilesFromPath(realPath, relativePath + '/' + p);
                results = results.concat(child);

            } else {
                const fileInfo = path.parse(realPath);

                if (fileInfo.ext === '.js') {
                    results.push({
                        filename: p,
                        realPath: realPath,
                        name: fileInfo.name,
                        relativePath: relativePath + '/' + fileInfo.name
                    });
                }
            }
        });

    }

    return results;
}

module.exports = function (basePath = path.resolve(__dirname, '../salesforce')) {
    // Create the base configuration
    const webpackConfig = {
        resolve: {
            alias: {}
        }
    }

    // Generate the alias
    const files = getFilesFromPath(basePath, '');
    files.forEach(f => {
        webpackConfig.resolve.alias['@salesforce' + f.relativePath] = f.realPath;
    });
    
    return webpackConfig;
}
