// @omnistudio/resources/src/modules/lightning/spinner/spinner.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v109.0.0
 */
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { LightningElement, api } from 'lwc';
import { classSet } from 'lightning/utils';
import { normalizeString as normalize } from 'lightning/utilsPrivate';

export default class cSpinner extends LightningElement {
    @api alternativeText;

    @api size = 'medium';

    @api variant;

    connectedCallback() {
        this.classList.add('slds-spinner_container');
        this.template.addEventListener('mousewheel', this.stopScrolling);
        this.template.addEventListener('touchmove', this.stopScrolling);
    }

    get normalizedVariant() {
        return normalize(this.variant, {
            fallbackValue: 'base',
            validValues: ['base', 'brand', 'inverse']
        });
    }

    get normalizedSize() {
        return normalize(this.size, {
            fallbackValue: 'medium',
            validValues: ['small', 'medium', 'large']
        });
    }

    get computedClass() {
        const { normalizedVariant, normalizedSize } = this;
        const classes = classSet('slds-spinner');

        if (normalizedVariant !== 'base') {
            classes.add(`slds-spinner_${normalizedVariant}`);
        }

        classes.add(`slds-spinner_${normalizedSize}`);

        return classes.toString();
    }

    get validAlternativeText() {
        const hasAlternativeText = !!this.alternativeText;

        if (!hasAlternativeText) {
            // eslint-disable-next-line no-console
            console.warn(
                `<lightning-spinner> The alternativeText attribute should not be empty. Please add a description of what is causing the wait.`
            );
        }

        return hasAlternativeText;
    }

    stopScrolling(event) {
        event.preventDefault();
    }
}
