// @omnistudio/resources/src/modules/lightning/layout/layout.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v109.0.0
 */
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { LightningElement, api, track } from 'lwc';
import {
    normalizeParam,
    computeLayoutClass,
    HORIZONTAL_ALIGN,
    VERTICAL_ALIGN,
    BOUNDARY
} from './styleUtils';
import { normalizeBoolean } from 'lightning/utilsPrivate';

export default class cLayout extends LightningElement {
    @api get horizontalAlign() {
        return this._horizontalAlign;
    }
    set horizontalAlign(value) {
        this._horizontalAlign = normalizeParam(value, HORIZONTAL_ALIGN);
        this.updateClassList();
    }
    @track _horizontalAlign;

    @api get verticalAlign() {
        return this._verticalAlign;
    }
    set verticalAlign(value) {
        this._verticalAlign = normalizeParam(value, VERTICAL_ALIGN);
        this.updateClassList();
    }
    @track _verticalAlign;

    @api get pullToBoundary() {
        return this._pullToBoundary;
    }
    set pullToBoundary(value) {
        this._pullToBoundary = normalizeParam(value, BOUNDARY);
        this.updateClassList();
    }
    @track _pullToBoundary;

    @api get multipleRows() {
        return this._multipleRows || false;
    }
    set multipleRows(value) {
        this._multipleRows = normalizeBoolean(value);
        this.updateClassList();
    }
    @track _multipleRows;

    _layoutClass = [];

    connectedCallback() {
        this.updateClassList();
    }

    updateClassList() {
        this.classList.remove(...this._layoutClass);
        const config = computeLayoutClass(
            this.horizontalAlign,
            this.verticalAlign,
            this.pullToBoundary,
            this.multipleRows
        );

        this._layoutClass = Object.keys(config);
        this.classList.add(...this._layoutClass);
    }
}
