// @omnistudio/components/src/modules/c/button/button.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { LightningElement, api, track } from "lwc";
import sldsTemplate from "./button_slds.html";
import ndsTemplate from "./button_nds.html";
import { lwcPropertyNameConversion } from "c/utility";

export default class Button extends LightningElement {
  @api label;
  @api ariaLabel;
  @api variant;
  @api title;
  @api theme = "slds";
  @api disabled;
  @api iconName;
  @api ndsIcon; // Deprecated Property
  @api size;
  @api bgColor;
  @api color;
  @api extraclass;
  @api iconSize;
  @api iconUrl;
  @api iconFill;
  @api iconBgColor;
  @api iconVariant;
  @api type = "button";
  @api disableTab;
  @track isIconRight;
  @track labelStyle = "";
  @track _iconPosition;

  static delegatesFocus = true;

  @api get iconPosition() {
    return this._iconPosition;
  }
  set iconPosition(val) {
    this.isIconRight = val === "right" ? true : false;
    this._iconPosition = val;
  }

  @api get styles() {
    return this._styles;
  }

  set styles(val) {
    const validObj = str => {
      try {
        return JSON.parse(str);
      } catch (e) {
        return {};
      }
    };
    val = val ? (typeof val === "string" ? validObj(val) : val) : {};
    this._styles = val;
    this.labelStyle = "";
    if (val && val.label) {
      let keys = Object.keys(val.label);
      keys.forEach(key => {
        if (val.label[key] && key !== "textAlign") {
          this.labelStyle += `${lwcPropertyNameConversion(key)}:${
            val.label[key]
          };`;
        }
      });
    }
  }

  _tooltipTabIndex;

  render() {
    if (this.theme === "nds") {
      return ndsTemplate;
    }
    return sldsTemplate;
  }

  connectedCallback() {
    this.isIconRight = this.iconPosition === "right" ? true : false;
    if (this.label && !this._iconPosition) this._iconPosition = "left";
    this._tooltipTabIndex = this.disableTab ? -1 : undefined;
    this.template.addEventListener("click", this.focusBtn);
  }
  triggerEvent = event => {
    if (event) {
      this.dispatchEvent(
        new CustomEvent(event.type, {
          bubbles: true,
          composed: true
        })
      );
    }
  };
  renderedCallback() {
    const btn = this.template.querySelector(".vlocity-btn");
    if (btn && !this.isRendered) {
      this.isRendered = true;
      ["blur", "focus"].forEach(event => {
        btn.addEventListener(event, this.triggerEvent);
      });
    }
    if (!this.isFirstRender && (this.bgColor || this.color)) {
      if (this.bgColor) {
        const buttonEl = this.template.querySelector("button");
        if (buttonEl) {
          buttonEl.style.backgroundColor = this.bgColor;
        }
      }
      if (this.color) {
        //const label = this.template.querySelector(".btnLabel");
        //if (label) label.style.color = this.color;
        this.labelStyle += `color:${this.color}`;
      }
      this.isFirstRender = true;
    }
  }
  disconnectedCallback() {
    const btn = this.template.querySelector(".vlocity-btn");
    if (btn) {
      ["blur", "focus"].forEach(event => {
        btn.removeEventListener(event, this.triggerEvent);
      });
    }
    this.template.removeEventListener("click", this.focusBtn);
  }
  focusBtn = () => {
    let btn = this.template.querySelector(".vlocity-btn");
    if (btn) {
      btn.focus();
    }
  };

  get themeClass() {
    const useTheme = this.theme ? this.theme : "slds";
    return (
      ` vlocity-btn ${useTheme}-button ` +
      (this.variant ? `${useTheme}-button_${this.variant} ` : "") +
      (this.extraclass ? this.extraclass : "") +
      (this.size ? ` ${useTheme}-m-around_${this.size}` : "")
    );
  }
}
