// @omnistudio/oui/src/modules/c/omniscriptValidation/omniscriptValidation.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { AggregatesValidation } from './aggregatesValidation';
import { HasValidation } from './hasValidation';

const VALIDATION_EVENTS = {
    INVALID: 'omniinvalid',
    VALID: 'omnivalid',
};

Object.freeze(VALIDATION_EVENTS);

function showHideValidityHook(newShow, comp) {
    if (newShow) {
        comp.checkValidity();
    } else if (comp._isValid === false && !comp._initialRender) {
        comp.dispatchOmniEventUtil(comp, { jsonPath: comp.jsonDef.lwcId }, VALIDATION_EVENTS.VALID);
        window.console.log('omniscriptValidation validation event - valid ' + comp.jsonDef.name);
    }
}

export { AggregatesValidation, HasValidation, VALIDATION_EVENTS, showHideValidityHook };
