// @omnistudio/cards/src/modules/c/carddesignerUtility/flexDesignerStyleUtils.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { getCardObjectFields } from "./getCardObjectFields";
export let FLEX_DESIGNER_STYLE_OBJECT = {
  theme: "",
  background: {
    color: "",
    image: "",
    size: "",
    repeat: "",
    position: ""
  },
  border: {
    type: "",
    width: "",
    color: "",
    radius: "",
    style: ""
  },
  padding: [],
  margin: [],
  text: {
    align: "",
    color: ""
  },
  height: "",
  minHeight: "",
  maxHeight: "",
  class: "",
  sizeClass: "",
  container: {
    class: ""
  },
  size: {
    isResponsive: false,
    default: "12"
  },
  selectedStyles: "",
  customClass: "",
  elementStyleProperties: {
    //Includes all element specific style
  },
  inlineStyle: ""
};

export let getDefaultStyleForElement = function(card, type) {
  let cardObjectFields = getCardObjectFields(card);
  if (card[cardObjectFields.StylingConfiguration]) {
    if (typeof card[cardObjectFields.StylingConfiguration] === "string") {
      card[cardObjectFields.StylingConfiguration] = JSON.parse(
        card[cardObjectFields.StylingConfiguration]
      );
    }
    let styleDefaults =
      card[cardObjectFields.StylingConfiguration].styleDefaults || {};
    let savedStyleObjects =
      card[cardObjectFields.StylingConfiguration].styleObjects || {};

    if (styleDefaults[type]) {
      return savedStyleObjects[styleDefaults[type]];
    }
  }
  return null;
};
