// @omnistudio/components/src/modules/c/pill/pill.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { LightningElement, api, track } from "lwc";
import sldsTemplate from "./pill_slds.html";
import ndsTemplate from "./pill_nds.html";
import pubsub from "c/pubsub";
import { pillLabels as translatedLabels } from "c/salesforceUtils";
import { lwcPropertyNameConversion } from "c/utility";

export default class Pill extends LightningElement {
  @api label;
  @api theme;
  @api disabled;
  @api extraclass = "";
  @api placeholder;
  @api freetext;
  @api viewonly;
  @api name;
  @api iconUrl;
  @api tagExtraclass = "";
  @api required;
  @track _options;
  @track inputValue;
  @track pillValue;
  @track islistvisible = false;
  @track valueMap = [];
  @track filteredOptions = [];

  @api
  get value() {
    return this.pillValue;
  }
  set value(value) {
    this.updatePill(value);
  }

  @api
  get options() {
    return this._options;
  }
  set options(value) {
    if (typeof value === "string") value = JSON.parse(value);
    this._options =
      value &&
      value.map(label => {
        return {
          id: this.uniqueId(),
          label: label
        };
      });
  }

  @track styleProperties = {
    label: {},
    value: {}
  };

  @api get styles() {
    return this._styles;
  }

  set styles(val) {
    const validObj = str => {
      try {
        return JSON.parse(str);
      } catch (e) {
        return {};
      }
    };

    val = val ? (typeof val === "string" ? validObj(val) : val) : {};
    val = val.styles ? val.styles : val;
    this._styles = val;
    if (this.styles) {
      for (let key in this.styles) {
        if (this._styles.hasOwnProperty(key)) {
          this.styleProperties[key] = {};
          if (key === "label") {
            this.styleProperties.label.styles = "";
            this.updateStyles(this.styles[key], key);
          } else if (key === "value") {
            this.updateStyles(this.styles[key], key);
          }
        }
      }
    }
  }

  updateStyles(styleObj, styleKey) {
    let keys = Object.keys(styleObj);
    keys.forEach(key => {
      if (!this.styleProperties[styleKey].styles) {
        this.styleProperties[styleKey].styles = "";
      }
      this.styleProperties[styleKey].styles += `${lwcPropertyNameConversion(
        key
      )}:${styleObj[key]};`;
    });
  }

  get pillClass() {
    return `${this.theme}-form-element ${this.extraclass}`;
  }

  updatePill(value) {
    this.pillValue = value;
    this.valueMap =
      value &&
      value.split(",").map(label => {
        return {
          id: this.uniqueId(),
          label: label
        };
      });
    this.valueMap = this.valueMap === "" ? [] : this.valueMap;
    this.filteredOptions = [];
  }

  render() {
    if (this.theme === "nds") {
      return ndsTemplate;
    }
    return sldsTemplate;
  }

  toggleLookup(event) {
    this.filterOptions();
    if (event.type === "focus" && this.filteredOptions.length > 0) {
      this.islistvisible = true;
    } else if (event.type === "blur") {
      this.islistvisible = false;
    }
  }
  fireEvent(eventName, data) {
    let event = new CustomEvent(eventName, {
      bubbles: true,
      composed: true,
      detail: {
        result: data || ""
      }
    });
    this.dispatchEvent(event);
  }

  enterInput(event) {
    if (this.freetext && event.keyCode === 13) {
      if (
        event.target.value.trim().length > 0 &&
        this.valueMap.indexOf(event.target.value) === -1
      ) {
        this.addValue(event.target.value);
        this.islistvisible = false;
      }
    }
  }
  selectOption(event) {
    this.inputValue = "";
    const item = event.currentTarget.getAttribute("data-field");
    this.addValue(item);
    this.fireEvent("select", event.currentTarget.dataset.field);
  }

  addValue(item) {
    this.valueMap = JSON.parse(JSON.stringify(this.valueMap));
    this.valueMap.push({
      id: this.uniqueId(),
      label: item
    });
    this.pillValue = this.valueMap.map(a => a.label).join(",");
    this.filterOptions();
    pubsub.fire(this.name, "valuechange", {
      name: this.name,
      value: this.pillValue
    });
  }

  filterOptions(event) {
    if (this.options) {
      this.filteredOptions = this.options.filter(option => {
        this.inputValue = event ? event.target.value : "";
        if (event && event.target.value) {
          return option.label
            .toLowerCase()
            .includes(event.target.value.toLowerCase());
        } else if (
          this.valueMap &&
          this.valueMap.map(a => a.label).indexOf(option.label) !== -1
        ) {
          return false;
        }
        return true;
      });
    }
  }

  translatedLabels = {};

  connectedCallback() {
    this.inputValue = "";
    this.translatedLabels = translatedLabels;
  }

  removeValue(event) {
    const item = event.currentTarget.getAttribute("data-field");
    this.valueMap.splice(this.valueMap.map(a => a.label).indexOf(item), 1);
    this.pillValue = this.valueMap.map(a => a.label).join(",");
    pubsub.fire(this.name, "valuechange", {
      name: this.name,
      value: this.pillValue
    });
    this.fireEvent("remove", event.currentTarget.dataset.field);
  }

  uniqueId() {
    return Date.now() + "-" + Math.random();
  }

  get itemTagClass() {
    let classes = `${this.tagExtraclass ? this.tagExtraclass : ""} ${
      this.theme
    }-listbox_selection-group ${this.theme}-m-vertical_x-small`;
    return classes;
  }
}
