// @omnistudio/resources/src/modules/lightning/formattedNumber/formattedNumber.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v109.0.0
 */
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { LightningElement, api } from 'lwc';
import { numberFormat } from 'lightning/internationalizationLibrary';

export default class cFormattedNumber extends LightningElement {
    @api value;

    @api formatStyle = 'decimal';

    @api currencyCode;

    @api currencyDisplayAs = 'symbol';

    @api minimumIntegerDigits;

    @api minimumFractionDigits;

    @api maximumFractionDigits;

    @api minimumSignificantDigits;

    @api maximumSignificantDigits;

    get formattedNumber() {
        const value = this.value;
        const options = {
            style: this.formatStyle,
            currency: this.currencyCode,
            currencyDisplay: this.currencyDisplayAs,
            minimumIntegerDigits: this.minimumIntegerDigits,
            minimumFractionDigits: this.minimumFractionDigits,
            maximumFractionDigits: this.maximumFractionDigits,
            minimumSignificantDigits: this.minimumSignificantDigits,
            maximumSignificantDigits: this.maximumSignificantDigits
        };

        const canReturnValue =
            value !== undefined &&
            value !== null &&
            value !== '' &&
            isFinite(value);

        if (canReturnValue) {
            const valueAsString = value.toString();

            let valueToFormat = valueAsString;

            if (this.formatStyle === 'percent-fixed') {
                options.style = 'percent';

                valueToFormat = parseFloat(value) / 100;

                if (
                    valueAsString.indexOf('.') > 0 &&
                    valueAsString.indexOf('e') < 0
                ) {
                    valueToFormat = valueToFormat.toFixed(
                        valueAsString.split('.')[1].length + 2
                    );
                }
            }
            return numberFormat(options).format(valueToFormat);
        }

        return '';
    }
}

cFormattedNumber.interopMap = {
    props: {
        formatStyle: 'style'
    }
};
