// @omnistudio/oui/src/modules/c/omniscriptDate/omniscriptDate.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import OmniscriptAtomicElement from 'c/omniscriptAtomicElement';
import { dateFormatCompatability, parseChronoAttributes } from 'c/omniscriptUtils';
import tmpl from './omniscriptDate.html';

/**
 * Omniscript Date Element
 */
export default class OmniscriptDate extends OmniscriptAtomicElement {
    _setValueStrict = false;

    handleBlur() {}

    handleChange(evt) {
        if (evt && evt.target) {
            this.applyCallResp(evt.target.value);
        }
    }

    get defaultValue() {
        let dfltValue = super.defaultValue;
        return parseChronoAttributes(dfltValue);
    }

    validateData(data) {
        var valid = data == null || data === '';
        var asDate, useAsDate;
        if (!valid) {
            if (this.jsonDef && this._propSetMap && this._propSetMap.dateType && this._propSetMap.dateType.toUpperCase() === 'STRING') {
                valid = typeof data === 'string';
            }
            if (!valid) {
                valid = data.constructor === Date || ((asDate = new Date(data)), (useAsDate = asDate.toString() !== 'Invalid Date'), useAsDate);
            }
            if (!valid) {
                asDate = parseChronoAttributes(data);
                if (asDate != null) {
                    useAsDate = true;
                    valid = true;
                }
            }
        }
        return { valid: valid, dataToApply: useAsDate ? asDate : data };
    }

    initCompVariables() {
        super.initCompVariables();
        this._messageWhenBadInput = this.allCustomLabelsUtil.OmniValidationBadInputDate;
        if (this.jsonDef && this._propSetMap) {
            if (!this._propSetMap.strictDateFormatting) {
                if (this._propSetMap.dateFormat) {
                    this._dateFormat = dateFormatCompatability(this._propSetMap.dateFormat);
                }
                if (this._propSetMap.modelDateFormat) {
                    this._modelDateFormat = dateFormatCompatability(this._propSetMap.modelDateFormat);
                }
            }
            if (this._propSetMap.minDate) {
                this._min = parseChronoAttributes(this._propSetMap.minDate, 'minDate');
            }
            if (this._propSetMap.maxDate) {
                this._max = parseChronoAttributes(this._propSetMap.maxDate, 'maxDate');
            }

            this._messageWhenRangeOverflow =
                String(this.allCustomLabelsUtil.OmniDateDisabledDay + ' ' + this.allCustomLabelsUtil.OmniDateMax) + ' %max%';
            this._messageWhenRangeUnderflow =
                String(this.allCustomLabelsUtil.OmniDateDisabledDay + ' ' + this.allCustomLabelsUtil.OmniDateMin) + ' %min%';
        }
    }

    setElementFormattedValue() {
        if (this.childInput) {
            const dispVal = this.childInput.displayValue;
            // odd structure to only call getter once;
            if (dispVal != null) {
                this._elementFormattedValue = dispVal;
            }
        }
        super.setElementFormattedValue();
    }

    render() {
        return tmpl;
    }
}
