// @omnistudio/oui/src/modules/c/omniscriptInternalUtils/notificationHelpers.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { ShowToastEvent } from 'lightning/platformShowToastEvent';

/**
 * Use this function to show a toast.
 * Utilizes the standard slds salesforce toast.
 * @param {object} config
 * @param {scope} comp
 */
export function showToast(config, comp) {
    const event = new ShowToastEvent({
        title: config.title,
        message: config.message,
        variant: config.variant || 'info',
        mode: config.mode || 'dismissable',
    });

    comp.dispatchEvent(event);
}
