// @omnistudio/components/src/modules/c/relatedList/relatedList.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { LightningElement, api, track } from "lwc";
import sldsTemplate from "./relatedList_slds.html";
import ndsTemplate from "./relatedList_nds.html";

let idGenerator = 0;
export default class RelatedList extends LightningElement {
  @api recordId;
  @api sObjectType;
  /*data table */
  @api issortavailable;
  @api extraclass;
  @api pagesize;
  @api issearchavailable;
  @api pagelimit;
  @api theme = "slds";
  @api iconUrl;
  @api hideTableHeader;
  @api groupBy;
  @api userSelectableColumn;
  @api groupOrder;
  @api cellLevelEdit;
  @api rowLevelEdit;
  @api rowDelete;
  @api draggable;
  @api records;
  @api columns;

  /*slds header */
  @api header;
  @api description;
  @api iconName;
  @api iconSize;
  @api iconColor;
  @api iconBgColor;
  @api imageRef;
  @api displayName;
  @api iconImgWrapperclass;

  @api actionList = [];
  @track actions = [];

  render() {
    if (this.theme === "nds") {
      return ndsTemplate;
    }
    return sldsTemplate;
  }

  renderedCallback() {
    if (!this.isRendered && this.actionList && this.actionList.length) {
      this.isRendered = true;
      // eslint-disable-next-line array-callback-return
      this.actionList.map(item => {
        this.actions.push({
          type: "Vlocity Action",
          name: item,
          id: "list-action-" + idGenerator++
        });
      });
    }
  }
}
