// @omnistudio/components/src/modules/c/accordionSection/accordionSection.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { LightningElement, api, track } from "lwc";
import sldsTemplate from "./accordionSection_slds.html";
import ndsTemplate from "./accordionSection_nds.html";

/**
 * @module c/accordion-section
 * @typicalname accordion-section
 * @example
 * ```
 * <c-accordion-section
 * label="Accordion With Menu"
 * name="A"
 * title="A"
 * theme="slds"
 * menuitem={menuitem}>
 * </c-accordion-section>
 *
 * ```
 */
export default class AccordionSection extends LightningElement {
  _uniqueKey;

  @track _iconName = "utility:chevronright";
  @track showactions = false;
  @track iconSize = "xx-small";
  @track menuSize = "xx-small";
  @track menuPosition = "right";

  /**
   * @type {String} - The name of the section. Needs to be unique inside one parent accordion
   * @scope api (public)
   */
  @api name = "";
  /**
   * @type {String} - The label of the section.
   * @scope api (public)
   */
  @api label;
  /**
   * @type {String} - The title of the section.
   * @scope api (public)
   */
  @api title = "";
  /**
   * @type {boolean} - If true, section will open and expanded. False by default.
   * @scope api (public)
   */
  @api isSectionOpen = false;

  //Deprecated
  @api extraClass = "";

  /**
   * @type {String} - Sets the theme of the component. Default is slds
   * @scope api (public)
   */
  @api get theme() {
    return this._theme;
  }
  set theme(val) {
    if (val === "slds" || val === "nds") {
      this._theme = val;
    } else {
      this._theme = "slds";
    }
  }

  /**
   * @type {String} - Set the icon name for the section arrow
   * @deprecated
   */
  @api get iconName() {
    return this._iconName;
  }
  set iconName(val) {
    this._iconName = val;
  }
  /**
   * @params {(String\|Array)} - Array of menu items
   * @scope api (public)
   */
  @api get menuitem() {
    return this._menuitem ? this._menuitem : "[]";
  }
  set menuitem(record) {
    this._menuitem = typeof record === "string" ? JSON.parse(record) : record;
  }

  get showMenu() {
    return this._menuitem && this._menuitem !== "[]" ? true : false;
  }

  get accordionIconName() {
    let iconName = this.isSectionOpen
      ? "utility:chevrondown"
      : "utility:chevronright";
    return iconName;
  }

  /**
   * @function - Dispatches a event with the section name and uniqueKey
   */
  toggleAccordion() {
    let event = new CustomEvent("togglesection", {
      bubbles: true,
      detail: { uniqueKey: this._uniqueKey, name: this.name } || ""
    });

    this.dispatchEvent(event);
  }

  /**
   * @function - Generates a unique key per section
   */
  uniqueKey() {
    return Date.now() + "-" + Math.random() + "-" + this.name;
  }

  get accordionClass() {
    let sectionOpenClass = this.isSectionOpen ? `${this.theme}-is-open` : "";
    return `${this.theme}-accordion__section ${sectionOpenClass}`;
  }

  connectedCallback() {
    this._uniqueKey = this.uniqueKey();
    this.classList.add(
      "accordion-section-class",
      `accordion-class_${this._uniqueKey}`
    );
  }

  render() {
    if (this.theme === "nds") {
      return ndsTemplate;
    }
    return sldsTemplate;
  }
}
