// @omnistudio/components/src/modules/c/img/img.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { LightningElement, api, track } from "lwc";
import sldsTemplate from "./img_slds.html";
import ndsTemplate from "./img_nds.html";
import { getSldsResourcesUrl } from "c/salesforceUtils";
import { newport, getCustomNewportUrl, isCommunityPage } from "c/utility";

export default class Img extends LightningElement {
  @api size;
  @api cropsize;
  @api theme = "slds";
  @api alternativeText = "";
  @api title;
  @api extraclass;
  @api titleclass;
  @api imgHeight;
  @api imgWidth;

  @track _imgsrc;

  @api get imgsrc() {
    return this._imgsrc;
  }
  set imgsrc(val) {
    this.isSrcRelative = !new RegExp("^(?:[a-z]+:)?//", "i").test(val);
    if (this.isSrcRelative) {
      this.srcVal = val;
      this._imgsrc =
        (this.isCommunity
          ? this.getCommunityPrefix()
          : window.location.origin) + val;
    } else {
      this._imgsrc = val;
    }
  }

  ndsUrl = "";
  _isCustomNdsUrlResolved = false;

  render() {
    if (this.theme === "nds") {
      return ndsTemplate;
    }
    return sldsTemplate;
  }

  connectedCallback() {
    isCommunityPage().then(result => {
      this.isCommunity = result;
      if (this.isCommunity && this.isSrcRelative && this.srcVal) {
        this._imgsrc = this.getCommunityPrefix() + this.srcVal;
      }
    });
    getCustomNewportUrl().then(url => {
      this._isCustomNdsUrlResolved = true;
      if (url) {
        this.ndsUrl = url.split("/assets")[0];
      }
    });
  }

  renderedCallback() {
    this.setImgDimension();
  }
  getCommunityPrefix() {
    if (/livepreview/.test(window.location.host)) {
      return "/sfsites/c";
    }
    return window.location.pathname.split("/s/")[0];
  }

  setImgDimension() {
    if (this.size) {
      this.imgHeight = null;
      this.imgWidth = null;
    }

    let imgElement = this.template.querySelector("img");
    if (imgElement) {
      imgElement.style = imgElement.style ? imgElement.style : {};

      if (this.imgHeight) {
        imgElement.style.height = this.imgHeight;
      } else {
        delete imgElement.style.height;
      }

      if (this.imgWidth) {
        imgElement.style.width = this.imgWidth;
      } else {
        delete imgElement.style.width;
      }
    }
  }

  get getClasses() {
    let classes = this.theme + "-image ";
    classes += this.extraclass ? this.extraclass : "";
    classes += this.size ? ` ${this.theme}-size_${this.size}` : "";
    classes += this.cropsize
      ? ` ${this.theme}-image__crop ${this.theme}-image__crop--${this.cropsize}`
      : "";
    return classes;
  }

  get getTitleClasses() {
    let classes = `${this.theme}-image__title ${this.theme}-image__title--card `;
    classes += this.titleclass ? this.titleclass : "";
    return classes;
  }

  get getAltIconUrl() {
    if (this.theme === "nds" && this._isCustomNdsUrlResolved) {
      if (this.ndsUrl)
        return (
          this.ndsUrl + "/assets/icons/doctype-sprite/svg/symbols.svg#image"
        );
      return newport + "/assets/icons/doctype-sprite/svg/symbols.svg#image";
    }
    return getSldsResourcesUrl() + "icons/doctype-sprite/svg/symbols.svg#image";
  }
}
