// @omnistudio/resources/src/modules/lightning/map/mapConstants.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v109.0.0
 */
export const DEFAULT_CENTER = { location: { Latitude: 0, Longitude: 180 } };
export const DEFAULT_ZOOM = 1;

export function lwcToGoogleLatLng(sfLatLang) {
  return new google.maps.LatLng(sfLatLang.Latitude, sfLatLang.Longitude);
}
/**
 *
 * @param {MapMarker[]} mapMarkers
 * @param {'Latitude' | 'Longitude'} property
 */
export function getMaxBounds(mapMarkers, property) {
  const maxVal = mapMarkers.reduce(
    // eslint-disable-next-line no-confusing-arrow
    (max, marker) =>
      marker.location[property] > max ? marker.location[property] : max,
    mapMarkers[0].location[property]
  );

  return Number(maxVal);
}

/**
 *
 * @param {MapMarker[]} mapMarkers
 * @param {'Latitude' | 'Longitude'} property
 */
export function getMinBounds(mapMarkers, property) {
  const minVal = mapMarkers.reduce(
    // eslint-disable-next-line no-confusing-arrow
    (min, marker) =>
      marker.location[property] < min ? marker.location[property] : min,
    mapMarkers[0].location[property]
  );

  return Number(minVal);
}

export const MAPS_NOT_AVAILABLE_MESSAGE =
  "Google Maps Platform was not found. Please include the the following in this page's header section.";

export const SCRIPT_EXAMPLE = `<script 
    type="text/javascript"
    src="https://maps.googleapis.com/maps/api/js?key={YOUR KEY}&libraries=places">
</script>`;

export const SCRIPT_REFERENCE =
  "https://developers.google.com/maps/documentation/javascript/examples/map-sync";
