// @omnistudio/components/src/modules/c/progressBar/progressBar.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { LightningElement, api } from "lwc";
import { progressBarLabels as translatedLabels } from "c/salesforceUtils";

/**
 * Custom element that will render a slds-progress-bar or nds-progress bar css component.
 * @module ns/progressBar
 * @typicalName ProgressBar
 * @extends LightningElement
 * @example
 * ``` html
 * <c-progress-bar progress="0"
 *            size="x-small"
 *            theme="nds"
 *            transition="0.5"
 *            success>
 * </c-progress-bar>
 * ```
 */
export default class ProgressBar extends LightningElement {
  //#region @api properties
  /**
   * @property {number} - Value representing the progress bar's completeness. Valid values from 0-100.
   * @scope api (public)
   * @default 0
   */
  @api get progress() {
    return this._progress;
  }
  set progress(value) {
    this.setProgress(value);
  }
  _progress = 0;

  /**
   * Corresponds to slds/nds styles determining size.
   * @property {'xx-small'|'x-small'|'small'|'medium'|'large'}
   * @default "small"
   * @scope api (public)
   */
  @api size = "small";

  /**
   * Number in seconds. Represents the css property transition-duration.
   * @property {number}
   * @scope api (public)
   */
  @api transition = 1;

  /**
   * Represents the theme being used. Either 'slds', or 'nds'.
   * @property {string}
   * @default "slds"
   * @scope api (public)
   */
  @api theme = "slds";

  /**
   * Flag when tru will render the component with success styling.
   * @property {boolean}
   * @scope api (public)
   */
  @api get success() {
    return this._success;
  }
  set success(value) {
    this.setSuccess(value);
  }
  _success;
  //#endregion @api properties

  /**
   * The HTML element which visually represents % complete. Set in connectedCallback.
   * @property {HTMLElement}
   * @scope private
   */
  _valueElement;

  _translatedLabels = {};

  //#endregion private properties

  //#region @api methods
  /**
   * Convenience method to sets the % of completeness of the progress bar value.
   * @param {number} value - Value representing the progress bar's completeness. Valid values from 0-100.
   * @scope api (public)
   * @example ``` JS
   * //..
   * const progressBar = this.template.querySelector('c-progress-bar');
   * //...
   * // A fictional event handler which will update the progress bar.
   * handleProgressEvent(evt) {
   *     progressBar.setProgress(evt.detail);
   * }
   * ```
   */
  @api setProgress(value, transition) {
    this._progress = value;
    if (this._valueElement) {
      this._valueElement.style.transitionDuration = transition
        ? `${transition}s`
        : `${this.transition}s`;
      this._valueElement.style.width = `${value}%`;
    }
  }

  /**
   * Convenience method to set the value of ProgressBar.success.
   * @property {boolean} success - If success is true, the progress indicator will change to the success color.
   * @scope api (public
   * @example ``` JS
   * // A fictional event handler which will update the progress bar, and set success styling when progress reaches 100!
   * handleProgressEvent(evt) {
   *     progressBar.setProgress(evt.detail);
   *     if (evt.detail === '100') {
   *         progressBar.setSuccess(true);
   *     }
   * }
   * ```
   */
  @api setSuccess(success) {
    this._success = Boolean(success) && success !== "false" && success !== "0";

    if (!this._valueElement) return;

    if (this._success === true) {
      this._valueElement.classList.add(
        `${this.theme}-progress-bar__value_success`
      );
    } else {
      this._valueElement.classList.remove(
        `${this.theme}-progress-bar__value_success`
      );
    }
  }
  //#endregion @api methods

  //#region lifecycle callbacks.
  connectedCallback() {
    this._translatedLabels = translatedLabels;
  }

  get componentClasses() {
    return `${this.theme}-progress-bar ${this.theme}-progress-bar_${this.size}`;
  }

  get valueClasses() {
    return `${this.theme}-progress-bar__value`;
  }

  get textClass() {
    return `${this.theme}-assistive-text`;
  }

  renderedCallback() {
    if (!this._initialRenderComplete) {
      this._valueElement = this.template.querySelector(
        "[data-progress-bar-value]"
      );

      this.setProgress(this._progress);

      if (this._success) {
        this.setSuccess(this._success);
      }
      this._initialRenderComplete = true;
    }
  }
  //#endregion
}
