// @omnistudio/cards/src/modules/c/flexToggle/flexToggle.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { api } from "lwc";
import Toggle from "c/toggle";
import { BaseFlexElementMixin } from "c/baseFlexElementMixin";

export default class FlexToggle extends BaseFlexElementMixin(Toggle) {
  @api get record() {
    return this._record;
  }
  set record(val) {
    this._record = val;
    if (this._rawFieldValues) {
      if (this.isMergeFieldVal(this._rawFieldValues.checked)) {
        this.updateValue(this._rawFieldValues.checked, "checked");
      }
      if (this.isMergeFieldVal(this._rawFieldValues.value)) {
        this.updateValue(this._rawFieldValues.value, "value");
      }
      if (this.isMergeFieldVal(this._rawFieldValues.disabled)) {
        this.disabled = this._rawFieldValues.disabled;
      }
    }
    this.triggerRender();
  }

  isMergeFieldVal(val) {
    return val && typeof val === "string" && val.indexOf("{") !== -1;
  }

  triggerRender() {
    if (this._rawFieldValues?.toggleEnabledLabel) {
      this._toggleEnabledLabel = this.interpolateValue(
        this._rawFieldValues.toggleEnabledLabel
      );
    }
    if (this._rawFieldValues?.toggleDisabledLabel) {
      this._toggleDisabledLabel = this.interpolateValue(
        this._rawFieldValues.toggleDisabledLabel
      );
    }
    if (this._rawFieldValues?.options) {
      this.options = this._rawFieldValues.options;
    }
  }
}
