// @omnistudio/components/src/modules/c/demolwcLabels/accordionLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpAccordion = "Accordion";
const cmpAccordionA = "Accordion A";
const cmpTitle1 = "Accordion Component";
const cmpLabel1 = "Accordion With Menu";
const cmpLabelA = "A";
const cmpMessage1 = "This is the first accordion";
const cmpDot = ".";
const cmpLabel2 = "Accordion Without Menu";
const cmpLabelB = "B";
const cmpMessage2 = "This is the second accordion";
const cmpLabelC = "C";
const cmpMessage3 = "This is the third accordion";
const cmpAccordionB = "Accordion B";
const cmpTitle2 = "Accordion Component";

export {
  cmpAccordion,
  cmpAccordionA,
  cmpTitle1,
  cmpLabel1,
  cmpLabelA,
  cmpMessage1,
  cmpDot,
  cmpLabel2,
  cmpLabelB,
  cmpMessage2,
  cmpLabelC,
  cmpMessage3,
  cmpAccordionB,
  cmpTitle2
};
