// @omnistudio/components/src/modules/c/demolwcLabels/alertLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const closeLabel = "Close";
const cmpAlert = "Alert";
const cmpSimpleMessage = "This is a simple message.";
const cmpComplex = "complex";
const cmpUsingSlots = "Using slots.";
const cmpAlertMessage = "messaging, use the default slot.";
const cmpOfflineAlertMessage = "Offline: For more ";
const cmpErrorAlertMessage = "Error: For more ";
const cmpWarningAlertMessage = "Warning: For more";
const cmpInfoAlertMessage = "Information: For more";

export {
  closeLabel,
  cmpAlert,
  cmpSimpleMessage,
  cmpComplex,
  cmpUsingSlots,
  cmpAlertMessage,
  cmpOfflineAlertMessage,
  cmpErrorAlertMessage,
  cmpWarningAlertMessage,
  cmpInfoAlertMessage
};
