// @omnistudio/components/src/modules/c/demolwcLabels/commonLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpVlocityLWCDemo = "Vlocity LWC Demo";
const cmpvVlocityTabComponent = "Vlocity Tab Component";
const cmpRadioGroups = "Radio Groups";
const cmpCheckboxGroups = "Checkbox Groups";
const cmpToggle = "Toggle";
const cmpRadioImageGroups = "Radio Image Groups";
const cmpCheckboxImageGroups = "Checkbox Image Groups";
const cmpRTE = "RTE";
const cmpRichTextEditor = "Rich Text Editor";
const cmpCharts = "Charts";
const cmpMonacoEditor = "Monaco Editor";
const cmpInteractionTracking = "InteractionTracking";
const cmpInteractiontrackingLabel = "Interaction Tracking";
const cmpEnterName = "Event Name";
const cmpTree = "Tree";

export {
  cmpVlocityLWCDemo,
  cmpvVlocityTabComponent,
  cmpRadioGroups,
  cmpCheckboxGroups,
  cmpToggle,
  cmpRadioImageGroups,
  cmpCheckboxImageGroups,
  cmpRTE,
  cmpRichTextEditor,
  cmpCharts,
  cmpMonacoEditor,
  cmpInteractionTracking,
  cmpInteractiontrackingLabel,
  cmpEnterName,
  cmpTree
};
