// @omnistudio/components/src/modules/c/demolwcLabels/dataTableLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpDataTable = "Data Table";
const cmpDataTableComponent = "Data Table Component";
const cmpTableWithCustomColumnId = "Table with custom Column UI";
const cmpEmail = "email";
const cmpAge = "age";
const cmpBalance = "balance";
const cmpDOB = "dateofbirth";
const cmpAddress = "address";
const cmpIsActive = "isActive";
const cmpPhone = "phone";
const cmpDataTableWithGrouping = "Data table with grouping";
const cmpDataTableWithoutGrouping = "Data table without grouping";

export {
  cmpDataTable,
  cmpDataTableComponent,
  cmpTableWithCustomColumnId,
  cmpEmail,
  cmpAge,
  cmpBalance,
  cmpDOB,
  cmpAddress,
  cmpIsActive,
  cmpPhone,
  cmpDataTableWithGrouping,
  cmpDataTableWithoutGrouping
};
