// @omnistudio/components/src/modules/c/demolwcLabels/datetimePickerLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpSelectDate = "Select Date";
const cmpSelectTime = "Select Time";
const cmpSelectDatetime = "Select Datetime";
const cmpFieldRequired = "This field is required.";
const cmpFieldNotValid = "This field is not valid. Please use the format {0}.";
const cmpDateFieldNotValid = "This field is not valid.";
const cmpFieldValueMissing = "This field value is missing.";
const cmpInvalidValue = "Invalid Value.";
const cmpValueBelowMinValue = "Value is below the min value.";
const cmpValueAboveMaxValue = "Value is above the max value.";
const cmpRequired = "required";
const cmpDate = "Date";
const cmpTime = "Time";
const cmpDayJsLocaleFormats =
  '{"name":"en","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],"months":["January","February","March","April","May","June","July","August","September","October","November","December"]}';
const cmpLocaleFormatsInvalid =
  'Unable to render ${1} due to invalid localeFormat [${2}]. Please provide a valid localeFormat in the form of {"name":"en","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],"months":["January","February","March","April","May","June","July","August","September","October","November","December"]}';
const cmpDateTimePicker = "Date/time Picker";
const cmpDataTimeDatetimePicker = "Date/Time/Datetime Picker";
const cmpDateInputDemo = "Date Input Demo";
const cmpDateLabel = "DATE";
const cmpDateValue1 = "20/03/2018";
const cmpDateMin1 = "10/03/2018";
const cmpDateFormat1 = "DD/MM/YYYY";
const cmpDateOutputFormat1 = "DD-MM-YYYY";
const cmpDateName1 =
  "Date Picker(DD/MM/YYYY format and DD-MM-YYYY output format) with min(2018-03-10)";
const cmpDateValue2 = "2019-03-28";
const cmpDateMin2 = "2019-03-10";
const cmpDateMax2 = "2019-04-10";
const cmpDateName2 = "Date Picker(YYYY-MM-DD format) with min and max values";
const cmpDateValue3 = "19-03-2019";
const cmpDateMax3 = "20-03-2019";
const cmpDateFormat3 = "DD-MM-YYYY";
const cmpDateName3 =
  "Date Picker(DD/MM/YYYY format and DD-MMM-YYYY output format) with max(2019-03-20)";
const cmpDateFieldHelp4 = "Help description.";
const cmpDateFormat4 = "MM/DD/YYYY";
const cmpDateOutputFormat4 = "DD/MM/YYYY";
const cmpDateValue4 = "2017-09-07T00:00:00Z";
const cmpDateName4 = "Date Picker(MM/DD/YYYY format)";
const cmpDateFormat5 = "DD-MMM-YYYY";
const cmpDateName5 = "Date Picker(DD-MMM-YYYY format)";
const cmpTimeInputDemo = "Time Input Demo";
const cmpTimeFieldHelp1 = "Time input";
const cmpTimeLabel1 = "Time (hh:mm a)";
const cmpTimeFormat1 = "hh:mm a";
const cmpTimeLabel2 = "Time (hh:mm:ss a)";
const cmpTimeValue2 = "08:10:00 AM";
const cmpTimeFormat2 = "hh:mm:ss a";
const cmpTimeLabewl3 = "Time (HH:mm:ss.SSS)";
const cmpTimeValue3 = "1:00:00.000";
const cmpTimeFormat3 = "HH:mm:ss.SSS";
const cmpTimeLabel4 = "Time (default format : HH:mm)";
const cmpTimeLabel5 = "Time (min 11:00)";
const cmpTimeMin5 = "11:00";
const cmpTimeLabel6 = "Time (max 13:00)";
const cmpTimeMax6 = "13:00";
const cmpTimeLabel7 = "Time (max: 14:00, min: 10:00)";
const cmpTimeMax7 = "14:00";
const cmpTimeMin7 = "10:00";
const cmpTimeLabel8 = "Time (interval : 10)";
const cmpTimeInterval8 = "10";
const cmpDatetimeInputDemo = "Datetime Input Demo";
const cmpDatetimELabel1 = "Select Datetime(formatted)";
const cmpDatetimeValue2 = "2019-03-30T10:15:00";
const cmpDatetimeLabel2 = "Select Datetime (empty date/time label)";
const cmpDatetimeLabel3 = "Select Datetime (output-type:date)";
const cmpDatetimeLabel4 = "Select Datetime (required)";
const cmpDatetimeValue5 = "2018-09-12T00:15:00";
const cmpDatetimeLabel5 = "Select Datetime (disabled)";
const cmpDatetimeMin6 = "2019-04-12T10:15:00";
const cmpDatetimeLabel6 = "Select Datetime (min:2019-04-12T10:15:00)";
const cmpDatetimeMax7 = "2019-05-12T18:15:00";
const cmpDatetimeLabel7 = "Select Datetime (max:2019-05-12T18:15:00)";

export {
  cmpSelectDate,
  cmpSelectTime,
  cmpSelectDatetime,
  cmpFieldRequired,
  cmpFieldNotValid,
  cmpDateFieldNotValid,
  cmpFieldValueMissing,
  cmpInvalidValue,
  cmpValueBelowMinValue,
  cmpValueAboveMaxValue,
  cmpRequired,
  cmpDate,
  cmpTime,
  cmpDayJsLocaleFormats,
  cmpLocaleFormatsInvalid,
  cmpDateTimePicker,
  cmpDataTimeDatetimePicker,
  cmpDateInputDemo,
  cmpDateLabel,
  cmpDateValue1,
  cmpDateMin1,
  cmpDateFormat1,
  cmpDateOutputFormat1,
  cmpDateName1,
  cmpDateValue2,
  cmpDateMin2,
  cmpDateMax2,
  cmpDateName2,
  cmpDateValue3,
  cmpDateMax3,
  cmpDateFormat3,
  cmpDateName3,
  cmpDateFieldHelp4,
  cmpDateFormat4,
  cmpDateOutputFormat4,
  cmpDateValue4,
  cmpDateName4,
  cmpDateFormat5,
  cmpDateName5,
  cmpTimeInputDemo,
  cmpTimeFieldHelp1,
  cmpTimeLabel1,
  cmpTimeFormat1,
  cmpTimeLabel2,
  cmpTimeValue2,
  cmpTimeFormat2,
  cmpTimeLabewl3,
  cmpTimeValue3,
  cmpTimeFormat3,
  cmpTimeLabel4,
  cmpTimeLabel5,
  cmpTimeMin5,
  cmpTimeLabel6,
  cmpTimeMax6,
  cmpTimeLabel7,
  cmpTimeMax7,
  cmpTimeMin7,
  cmpTimeLabel8,
  cmpTimeInterval8,
  cmpDatetimeInputDemo,
  cmpDatetimELabel1,
  cmpDatetimeValue2,
  cmpDatetimeLabel2,
  cmpDatetimeLabel3,
  cmpDatetimeLabel4,
  cmpDatetimeValue5,
  cmpDatetimeLabel5,
  cmpDatetimeMin6,
  cmpDatetimeLabel6,
  cmpDatetimeMax7,
  cmpDatetimeLabel7
};
