// @omnistudio/components/src/modules/c/demolwcLabels/demolwcLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import * as alertLabels from "./alertLabels";
import * as checkboxGroupLabels from "./checkboxGroupLabels";
import * as componentConstraintApiLabels from "./componentConstraintApiLabels";
import * as inputLabels from "./inputLabels";
import * as pillLabels from "./pillLabels";
import * as progressBarLabels from "./progressBarLabels";
import * as textareaLabels from "./textareaLabels";
import * as typeaheadLabels from "./typeaheadLabels";
import * as datePickerLabels from "./datePickerLabels";
import * as datetimePickerLabels from "./datetimePickerLabels";
import * as timePickerLabels from "./timePickerLabels";
import * as commonLabels from "./timePickerLabels";
import * as listLabels from "./listLabels";
import * as menuLabels from "./menuLabels";
import * as actionLabels from "./actionLabels";
import * as formLabels from "./formLabels";
import * as modalLabels from "./modalLabels";
import * as promptLabels from "./promptLabels";
import * as popoverLabels from "./popoverLabels";
import * as toastLabels from "./toastLabels";
import * as relatedListLabels from "./relatedListLabels";
import * as dataTableLabels from "./dataTableLabels";
import * as carouselLabels from "./carouselLabels";
import * as layoutLabels from "./layoutLabels";
import * as sliderLabels from "./sliderLabels";
import * as tooltipLabels from "./tooltipLabels";
import * as customLabels from "./customLabels";
import * as propertyPanelLabels from "./propertyPanelLabels";
import * as accordionLabels from "./accordionLabels";
import * as sldsHeaderLabels from "./sldsHeaderLabels";
import * as wizardLabels from "./wizardLabels";

export {
  alertLabels,
  checkboxGroupLabels,
  componentConstraintApiLabels,
  datePickerLabels,
  datetimePickerLabels,
  inputLabels,
  pillLabels,
  progressBarLabels,
  textareaLabels,
  typeaheadLabels,
  timePickerLabels,
  commonLabels,
  listLabels,
  menuLabels,
  actionLabels,
  formLabels,
  modalLabels,
  promptLabels,
  popoverLabels,
  toastLabels,
  relatedListLabels,
  dataTableLabels,
  carouselLabels,
  layoutLabels,
  sliderLabels,
  tooltipLabels,
  customLabels,
  propertyPanelLabels,
  accordionLabels,
  sldsHeaderLabels,
  wizardLabels,
};
