// @omnistudio/components/src/modules/c/demolwcLabels/formLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpForm = "Form";
const cmpFormComponent = "Form Component";
const cmpName = "Name";
const cmpPassword = "Password";
const cmpTypeahead = "Typeahead";
const cmpCombobox = "Combobox";
const cmpNormalInput = "Normal Input";
const cmpHelpDescription = "Help description.";
const cmpNormalCheckbox = "Normal Checkbox";
const cmpYes = "Yes";
const cmpNo = "No";

export {
  cmpForm,
  cmpFormComponent,
  cmpName,
  cmpPassword,
  cmpTypeahead,
  cmpCombobox,
  cmpNormalInput,
  cmpHelpDescription,
  cmpNormalCheckbox,
  cmpYes,
  cmpNo
};
