// @omnistudio/components/src/modules/c/demolwcLabels/propertyPanelLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpPropertyPanel = "Property Panel";
const cmpPropertyPanelLabel = "PROPERTY PANEL";
const cmpTitle1 = "collapsible section 1";
const cmpTitle2 = "collapsible section 2";
const cmpTitle3 = "additem";
const cmpDone = "Done";
const cmpCancel = "Cancel";
const cmpLabel1 = "Open Property Panel";

export {
  cmpPropertyPanel,
  cmpPropertyPanelLabel,
  cmpTitle1,
  cmpTitle2,
  cmpTitle3,
  cmpDone,
  cmpCancel,
  cmpLabel1
};
