// @omnistudio/components/src/modules/c/demolwcLabels/sldsHeaderLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpSldsHeader = "Slds Header";
const cmpSampleheader = "Sample Header";
const cmpSampleHeaderWithBaseField = "Slds header with base fields";
const cmpCustombutton1 = "Custom Button 1";
const cmpCustombutton2 = "Custom Button 2";
const cmpCustombutton3 = "Custom Button 3";
const cmpMoreActions = "More Actions";
const cmpSampleHeader = "Sample Header";
const cmpSldsHeaderWithCustomFields = "Slds header with custom fields";
const cmpTitle1 = "Field 1";
const cmpTitle2 =
  "Description that demonstrates truncation with a long text field.";
const cmpTitle3 = "Field 2 (3)";
const cmpTitle4 = "Multiple Values";
const cmpTitle5 = "Field 3";
const cmpTitle6 = "Hyperlink";
const cmpTitle7 = "Field 4";
const cmpTitle8 = "Description (2-line truncation—must use JS to truncate).";
const cmpTitle9 = "Description (2-line truncati...";
const cmpSldsHeaderWithBaseFieldsAndMetaText =
  "Slds header with base fields and meta text";
const cmpMessage1 = "10 items • Updated 13 minutes ago";
const cmpSldsHeaderWithNoFields = "Slds header with no fields";
const cmpClickToEdit = "Click to Edit !";
const cmpClickToSave = "Click to Save !";

export {
  cmpSldsHeader,
  cmpSampleheader,
  cmpSampleHeaderWithBaseField,
  cmpCustombutton1,
  cmpCustombutton2,
  cmpCustombutton3,
  cmpMoreActions,
  cmpSampleHeader,
  cmpSldsHeaderWithCustomFields,
  cmpTitle1,
  cmpTitle2,
  cmpTitle3,
  cmpTitle4,
  cmpTitle5,
  cmpTitle6,
  cmpTitle7,
  cmpTitle8,
  cmpTitle9,
  cmpSldsHeaderWithBaseFieldsAndMetaText,
  cmpMessage1,
  cmpSldsHeaderWithNoFields,
  cmpClickToEdit,
  cmpClickToSave
};
