// @omnistudio/components/src/modules/c/demolwcLabels/sliderLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpSlider = "Slider";
const cmpMinValue = "9000";
const cmpMaxValue = "12000";
const cmpToggles = "Toggles";
const cmpTogglePosition = "Toggle Position";
const cmpDisableEnable = "Disable/Enable";
const cmpWithStep = "With Step";
const cmpSizes = "Sizes";
const cmpXSmall = "X-Small";
const cmpSmall = "Small";
const cmpMedium = "Medium";
const cmpLarge = "Large";

export {
  cmpSlider,
  cmpMinValue,
  cmpMaxValue,
  cmpToggles,
  cmpTogglePosition,
  cmpDisableEnable,
  cmpWithStep,
  cmpSizes,
  cmpXSmall,
  cmpSmall,
  cmpMedium,
  cmpLarge
};
