// @omnistudio/components/src/modules/c/demolwcLabels/textareaLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpBadInput = "Bad input.";
const cmpValueTooShort = "Value is too short.";
const cmpValueTooLong = "Value is too long.";
const cmpFieldRequired = "This field is required.";
const cmpRequired = "Required";
const cmpValue = "Value";
const cmpReadonly = "ReadOnly";
const cmpDisabled = "Disabled";
const cmpTextarea = "Textarea";
const cmpTextareaElement = "Textarea Element";
const cmpVlocityTextareaElement = "Vlocity Textarea Element";
const cmpLabel1 = "Description";
const cmpFieldLevelHelp1 = "Help description.";
const cmpPlaceholder1 = "Enter Description...";
const cmpPlaceholder2 = "No label variant";
const cmpValue1 = "Some text";
const cmpValue2 = "Some text only read access";
const cmpLabel2 = "Message When Value Missing";
const cmpLabel3 = "Message When Value Short(30 char)";
const cmpLabel4 = "Message When Value Long(30 char)";

export {
  cmpBadInput,
  cmpValueTooShort,
  cmpValueTooLong,
  cmpFieldRequired,
  cmpRequired,
  cmpValue,
  cmpReadonly,
  cmpDisabled,
  cmpTextarea,
  cmpTextareaElement,
  cmpVlocityTextareaElement,
  cmpLabel1,
  cmpFieldLevelHelp1,
  cmpPlaceholder1,
  cmpPlaceholder2,
  cmpValue1,
  cmpValue2,
  cmpLabel2,
  cmpLabel3,
  cmpLabel4
};
