// @omnistudio/components/src/modules/c/demolwcLabels/toastLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpToast = "Toast";
const cmpToastComponent = "Toast Component";
const cmpSLDSnNDSComponent = "SLDS and NDS TOAST COMPONENT";
const cmpOpenErrorToast = "Open Error Toast";
const cmpOpenInfoToast = "Open Info Toast";
const cmpOpenSuccessToast = "Open Success Toast";
const cmpOpenWarningToast = "Open Warning Toast";
const cmpSLDS = "SLDS";
const cmpErrormessage = "This is a error toast";
const cmpSave = "Save";
const cmpInfomessage = "This is info toast";
const cmpSuccessMessage = "This is Success toast";
const cmpWarningMessage = "This is Warning toast";

export {
  cmpToast,
  cmpToastComponent,
  cmpSLDSnNDSComponent,
  cmpOpenErrorToast,
  cmpOpenInfoToast,
  cmpOpenSuccessToast,
  cmpOpenWarningToast,
  cmpSLDS,
  cmpErrormessage,
  cmpSave,
  cmpInfomessage,
  cmpSuccessMessage,
  cmpWarningMessage
};
