// @omnistudio/components/src/modules/c/demolwcLabels/tooltipLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpTooltip = "Tooltip";
const cmpHelptextComponent = "Helptext componennt";
const cmpLabel1 = "Arrow Position - Top Left";
const cmpContent1 = "This is a demo tooltip";
const cmpLabel2 = "Arrow Position - Top Right With Custom Markup";
const cmpContent2 = "This is a demo helptextsssss";
const cmpMessage1 = "This is Custom markup";
const cmpMessage2 = "You can put your markup inside tooltip";
const cmpLabel3 = "This is a Button";
const cmpLabel4 = "Arrow Position - Bottom Left";
const cmpContent4 = "This is a demo helptext";
const cmplabel5 = "Arrow Position - Bottom Right";
const cmpContent5 = "This is a demo helptext";

export {
  cmpTooltip,
  cmpHelptextComponent,
  cmpLabel1,
  cmpContent1,
  cmpLabel2,
  cmpContent2,
  cmpMessage1,
  cmpMessage2,
  cmpLabel3,
  cmpLabel4,
  cmpContent4,
  cmplabel5,
  cmpContent5
};
