// @omnistudio/components/src/modules/c/demolwcLabels/wizardLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpWizard = "Wizard";
const cmpStep1 = "Step1";
const cmpStep1Label = "STEP 1";
const cmpStep2 = "Step2";
const cmpStep2Label = "STEP 2";
const cmpStep3 = "Step3";
const cmpStep3Label = "STEP 3";
const cmpStep4 = "Step4";
const cmpStep4Label = "STEP 4";
const cmpStep5 = "Step5";
const cmpStep5Label = "STEP 5";
const cmpMinValue = "9000";
const cmpMaxValue = "12000";
const cmpLabel1 = "Name";
const cmpPlaceholder1 = "Normal Input";
const cmpLabel2 = "Password";
const cmpHelpText = "Help description.";
const cmpLabel3 = "Combobox";
const cmpLabel4 = "Normal Checkbox";
const cmpYes = "Yes";
const cmpNo = "No";
const cmpLabel5 = "Tag";
const cmpPlaceholder5 = "Please enter tag";
const cmpLabel6 = "Show";
const cmpSave = "Save";

export {
  cmpWizard,
  cmpStep1,
  cmpStep1Label,
  cmpStep2,
  cmpStep2Label,
  cmpStep3,
  cmpStep3Label,
  cmpStep4,
  cmpStep4Label,
  cmpStep5,
  cmpStep5Label,
  cmpMinValue,
  cmpMaxValue,
  cmpLabel1,
  cmpPlaceholder1,
  cmpLabel2,
  cmpHelpText,
  cmpLabel3,
  cmpLabel4,
  cmpYes,
  cmpNo,
  cmpLabel5,
  cmpPlaceholder5,
  cmpLabel6,
  cmpSave
};
