// @omnistudio/oui/src/modules/c/omniscriptNumber/omniscriptNumber.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
/**
 * @module ns/omniscriptNumber
 * @description This component is used to render a Number Element, OmniscriptNumber is extended from `OmniscriptAtomicElement`. Supported masking format example ##,##.##
 */
import OmniscriptAtomicElement from 'c/omniscriptAtomicElement';
import tmpl from './omniscriptNumber.html';

/**
 * Default exported class OmniscriptNumber.
 * @extends OmniscriptAtomicElement
 * @typicalname OmniscriptNumber
 */
export default class OmniscriptNumber extends OmniscriptAtomicElement {
    /**
     * @type {Object} - Hold an object from Helper method for mask handling
     * @scope private
     */
    _imaskNumberAttributes;

    /**
     * @type {Boolean} - Checks whether value is committed or not
     * @scope private
     */
    _commitOnChange;

    /**
     * @description Overwrites inherited initCompVariables.(this method executes once during connectedCallback.)
     * @returns {Void}
     * @scope private
     */
    initCompVariables() {
        super.initCompVariables();
        this._imaskNumberAttributes = this.getImaskNumberAttributes();
        this._commitOnChange = this.scriptHeaderDef.propSetMap.commitOnChange === true;
    }

    /**
     * Event handler function on blur event.
     * Sets the element value and triggers aggregation
     * @returns {void}
     * @scope private
     */
    handleBlur(evt) {
        this.applyCallResp(evt.target.value);
    }

    /**
     * @description Evaluates if currency is valid.
     * @param {Number} data
     * @returns {Object}
     * @scope private
     */
    validateData(data) {
        // default - Number
        return { valid: typeof data === 'number' || data === null, dataToApply: data };
    }

    /**
     * @scope private
     * @description Overwrites the native LWC render.
     * @returns {Template}
     */
    render() {
        return tmpl;
    }
}
