// @omnistudio/components/src/modules/c/propertySection/propertySection.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import { LightningElement, api, track } from "lwc";
export default class VlocityPropertySection extends LightningElement {
  @api title;
  @api isCollapsible;
  @track showSectionContent;
  @api isExpanded;
  @api showCount = false;
  @api count = 0;

  @track _isExpanded;
  get isExpanded() {
    return this._isExpanded;
  }
  set isExpanded(expanded) {
    this._isExpanded = expanded === "true" || expanded === true;
    this.showSectionContent = this.isExpanded;
  }

  renderedCallback() {
    if (this.showSectionContent) {
      this.template
        .querySelector(".section-content-container")
        .classList.add("open");
    }
  }

  expandSection() {
    this.showSectionContent = true;
    this.template
      .querySelector(".section-content-container")
      .classList.add("open");
  }

  collapseSection() {
    this.showSectionContent = false;
    this.template
      .querySelector(".section-content-container")
      .classList.remove("open");
  }

  get showExpandButton() {
    return this.isCollapsible && !this.showSectionContent;
  }

  get showCollapseButton() {
    return this.isCollapsible && this.showSectionContent;
  }

  get countDisplay() {
    return Number.isInteger(this.count) ? ` (${this.count}) ` : "";
  }
}
