// @omnistudio/oui/src/modules/c/omniscriptRestApi/customLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
export const allCustomLabels = {
    OmniSideBarError1: 'There are Actions embedded between Steps, therefore you can not navigate to a future Step.',
    OmniDesInvalidJson: 'The JSON is invalid',
    OmniLwcApexHttpActionNotAvailable: "Endpoint 'Apex' for HTTP Action is currently not available in Lightning Web Components.",
    OmniRemoteCallTimeout: 'Remote Call Timeout',
    OmniMaxInt: 'Maximum Permitted Value: 9007199254740991',
    OmniStepValidationError: 'Please fix all the fields with errors.',
    OmniValidationBadInput: 'Enter a valid value.',
    OmniValidationBadInputDate: 'Enter a valid value. Please use the format {0}.',
    OmniRequired: 'Required',
    OmniValidationPatternMismatch: 'Your entry does not match the allowed pattern.',
    OmniValidationRangeOverflow: 'The number is too high.',
    OmniValidationRangeUnderflow: 'The number is too low.',
    OmniValidationStepMismatch: "Your entry isn't a valid increment.",
    OmniMinLength: 'Minimum length of',
    OmniMaxLength: 'Maximum length of',
    OmniValidationTypeMismatch: 'You have entered an invalid format.',
    OmniValidationMaskIncomplete: 'This field is incomplete.',
    OmniDesOk: 'Ok',
    OmniRadioGroupSetAll: 'Set All',
    OmniInvalidURLError: 'Please enter a valid website URL. (e.g. https://www.example.com)',
    OmniInvalidEmailError: 'Invalid email format.',
    OmniScriptWebCalloutFailed: 'Web callout failed:',
    OmniScriptError: 'Error',
    OmniScriptNotFound1: 'There is no active OmniScript with:',
    OmniScriptType: 'Type',
    OmniScriptSubType: 'Sub Type',
    OmniScriptLang: 'Language',
    OmniScriptNotFound2: 'Please contact your Salesforce Administrator.',
    OmniScriptNotFound3: 'This OmniScript is either inactive or has been replaced with a new version which has not been deployed.',
    OmniDateMin: 'Please select a date on or after',
    OmniDateMax: 'Please select a date on or before',
    OmniTimeMinMax: 'Please select a time between',
    OmniDateDisabledDay: 'This date has been disabled.',
    OmnisaveLabel: 'Save for later',
    OmniSaved: 'Your OmniScript is saved for later',
    OmniSavedFailed:
        'Save for Later failed. If your form includes large attachments, please remove them and try again. Otherwise, navigating away from this page will lose all current information.',
    OmniSavedFailedConcurrent:
        'This saved OmniScript has been updated since resuming. To see the latest updates, please exit and resume the saved OmniScript again.',
    OmniSaveEmailBody: 'Note: for an authenticated user, please log in first before accessing the link.',
    OmniResume: 'To resume, please click or bookmark the following link',
    OmniLink: 'link',
    OmniCopyLink: 'Copy the link',
    OmniCopyToClipBoard2: 'Copy to clipboard: Double click to select, Ctrl+C/Cmd+C',
    OmniEmailMe: 'Email me the link',
    OmniResumeLink: 'Resume link:',
    OmniAdd: 'Add',
    OmniDelete: 'Delete',
    OmniScriptResumeLink: 'This script has been automatically saved, in order to resume in the future:',
    OmnisaveMessage: 'Are you sure you want to save it for later?',
    OmniConfirm: 'Confirm',
    OmniInvalidLwcComponentMessage: 'The form has been updated since you last saved it. Click Continue to start over using the updated form.',
    OmniInvalidLwcComponent: 'Form Updated',
    OmniContinue: 'Continue',
    OmnicancelLabel: 'Cancel',
    OmnicancelHeader: 'Cancel',
    OmnicancelMessage: 'Are you sure? Any unsaved changes will be lost.',
    OmniPrompt: 'Prompt',
    OmniEditBlockDeleteConfirmation: 'Are you sure that you want to permanently delete the selected item',
    OmniConfigProdModalOk: 'Save',
    OmniNavigationError1: 'This type of navigation is not supported.',
    OmniScriptOpen: 'Open in OmniScript',
    OmniScriptModal: 'Open in modal dialog',
    OmniScriptLastPublished: 'Last Published',
    OmniSendEmailNoTemplate: 'No Template found in the request',
    OmniDocuSignNoRecipients: 'No recipients were found in the request',
    OmniDocuSignModalTitle: 'DocuSign Signing Ceremony',
    OmniDocuSignModalViewPdf: 'View PDF',
    OmniDocuSignModalClose: 'Close',
    OmnipostMessage: 'Done',
    OmniPopupBlocked: 'Popup Blocked. Please disable any Popup Blocker and try again.',
    OmniBlockError: 'Please expand to fix all invalid fields.',
    OmniPendingUpdatesError: 'Please save or cancel any pending updates.',
    OmniOK: 'OK',
    OmniNoActionsError: 'No actions are defined.',
    OmnipreviousLabel: 'Previous',
    OmninextLabel: 'Next',
    OmnicompleteLabel: 'Complete',
    OmnisubmitLabel: 'Submit',
    OmnisummaryLabel: 'Summary',
    OmnireviseLabel: 'Go back',
    OmnifailureNextLabel: 'Continue',
    OmniconsoleTabLabel: 'New',
    OmninewItemLabel: 'New',
    OmninewLabel: 'New',
    OmnieditLabel: 'Edit',
    OmnicompleteMessage: 'Are you sure you want to complete the script?',
    OmniinProgressMessage: 'In Progress',
    OmniremoteActionConfirm: 'Are you sure? This action cannot be undone.',
    Omnicancel: 'Cancel',
    OmnifailureGoBackLabel: 'Go Back',
    OmniCustomLabelError: 'Failed to get custom labels',
    OmniDayJsLocaleFormats:
        '{"name":"en","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],"months":["January","February","March","April","May","June","July","August","September","October","November","December"]}',
    OmniLocaleFormatsInvalid:
        'Unable to render ${1} due to invalid localeFormat [${2}]. Please provide a valid localeFormat in the form of {"name":"en","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],"months":["January","February","March","April","May","June","July","August","September","October","November","December"]}',
    OmniToday: 'Today',
    OmniNextMonth: 'Next Month',
    OmniPrevMonth: 'Previous Month',
    OmniPickYr: 'Pick a year',
    OmniSelectDate: 'Select Date',
    OmniReturn: 'Return',
    ODCActionBlockDrop4ElementsBelow: 'Drop up to 4 child actions below:',
    OmniFileUploadCount: 'Number of files uploaded is {0}',
    OmniDeleteItem: 'Delete {0}',
    OmniTextRepeatAltText: 'Save {0} and add a new field',
    OmniTextRemoveAltText: 'Delete {0} and remove the field',
    OmniStepsNextStepAriaLabel: 'Next Step',
    OmniStepsPrevStepAriaLabel: 'Previous Step',
    OmniStepChartLabel: 'Steps',
    OmniStepHideDetails: 'Hide Details',
    OmniStepShowDetails: 'Show Details',
    OmniStepCloseDetails: 'Close Details',
    OmniSavingAccessibleLabel: 'Saving changes',
    OmniEditBlockItemCountAssistiveText: '{0} item',
    OmniEditBlockItemsCountAssistiveText: '{0} items',
    OmniCurrencyRepeatAltText: 'Save {0} and add a new currency field',
    OmniCurrencyRemoveAltText: 'Delete {0} and remove the currency field',
    OmniTypeAheadEditLabel: 'Edit',
    OmniKnowledgeSearchLabel: 'Search Knowledge',
    OmniKnowledgeSearchIcon: 'Search Icon',
    OmniKnowledgeExpand: 'Expand Knowledge',
    OmniKnowledgeCollapse: 'Collapse Knowledge',
    OmniBlockAddAriaLabel: 'Add a new block',
    OmniBlockRemoveAriaLabel: 'Delete block {0}',
    OmniEditBlockIconTextAdd: 'Add',
    OmniEditBlockIconTextUser: 'User',
    OmniEditBlockIconTextDown: 'Show More',
    OmniEditBlockIconTextApproval: 'Check',
    OmniEditBlockIconTextSave: 'Save',
    OmniEditBlockIconTextCancel: 'Cancel',
    OmniEditBlockIconTextDelete: 'Delete',
    OmniSpinnerTextLoading: 'Loading...',
    OmniLinkSuccess: 'Copied link to clipboard.',
    OmniCopyLinkError: 'Unable to copy link.',
    OmniScriptOpenFormatted: 'Open "{0}" in OmniScript',
    OmniScriptModalFormatted: 'Open "{0}" in modal dialog',
    OmniKnowledgeArticlesCount: '{0} articles found',
};

export function getLocalTranslations(language) {
    try {
        const translation = require('vlocitytranslations/translations')[language];
        return translation;
    } catch (error) {
        return {};
    }
}
