// @omnistudio/components/src/modules/c/salesforceUtils/dataTableLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpCancel = "Cancel";
const cmpDelete = "Delete";
const cmpDeleteConfirmationPrompt =
  "Are you sure you want to delete this item?";
const cmpLoading = "loading";
const cmpNext = "Next";
const cmpNoRecordsToDisplay = "No records to display";
const cmpPrevious = "Previous";
const cmpSearch = "Search";
const cmpSearchKey = "Search Key";
const cmpSortBy = "Sort By";

export {
  cmpSearchKey,
  cmpLoading,
  cmpSortBy,
  cmpPrevious,
  cmpNext,
  cmpNoRecordsToDisplay,
  cmpSearch,
  cmpDeleteConfirmationPrompt,
  cmpCancel,
  cmpDelete
};
