// @omnistudio/components/src/modules/c/salesforceUtils/datetimePickerLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpSelectDate = "Select Date";
const cmpSelectTime = "Select Time";
const cmpFieldRequired = "This field is required.";
const cmpFieldNotValid = "This field is not valid.";
const cmpDateFieldNotValid =
  "This field is not valid. Please use the format {0}.";
const cmpFieldValueMissing = "This field value is missing.";
const cmpInvalidValue = "Invalid Value.";
const cmpValueBelowMinValue = "Value is below the min value.";
const cmpValueAboveMaxValue = "Value is above the max value.";
const cmpRequired = "required";
const cmpDate = "Date";
const cmpTime = "Time";
const cmpDayJsLocaleFormats =
  '{"name":"en","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],"months":["January","February","March","April","May","June","July","August","September","October","November","December"]}';
const cmpLocaleFormatsInvalid =
  'Unable to render ${1} due to invalid localeFormat [${2}]. Please provide a valid localeFormat in the form of {"name":"en","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],"months":["January","February","March","April","May","June","July","August","September","October","November","December"]}';

export {
  cmpSelectDate,
  cmpSelectTime,
  cmpFieldRequired,
  cmpFieldNotValid,
  cmpDateFieldNotValid,
  cmpFieldValueMissing,
  cmpInvalidValue,
  cmpValueBelowMinValue,
  cmpValueAboveMaxValue,
  cmpRequired,
  cmpDate,
  cmpTime,
  cmpDayJsLocaleFormats,
  cmpLocaleFormatsInvalid
};
