// @omnistudio/components/src/modules/c/salesforceUtils/inputLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const cmpFieldNotValid = "This field is not valid.";
const cmpDateFieldNotValid =
  "This field is not valid. Please use the format {0}.";
const cmpFieldValueMissing = "This field value is missing.";
const cmpPatternNoMatch = "The pattern does not match.";
const cmpRangeOverflowError = "Range overflow error.";
const cmpRangeUnderflowError = "Range underflow error.";
const cmpStepMismatchError = "Step mismatch error.";
const cmpFieldTooShort = "This field is too short.";
const cmpFieldTooLong = "This field is too long.";
const cmpTypeMismatchError = "Type mismatch error.";

const cmpSelectDate = "Select Date";
const cmpSelectTime = "Select Time";
const cmpRequired = "required";

const cmpUpload = "Upload";
const cmpDelete = "Delete";
const cmpDate = "Date";
const cmpTime = "Time";

export {
  cmpFieldNotValid,
  cmpDateFieldNotValid,
  cmpFieldValueMissing,
  cmpPatternNoMatch,
  cmpRangeOverflowError,
  cmpRangeUnderflowError,
  cmpStepMismatchError,
  cmpFieldTooShort,
  cmpFieldTooLong,
  cmpTypeMismatchError,
  cmpSelectDate,
  cmpSelectTime,
  cmpRequired,
  cmpUpload,
  cmpDelete,
  cmpDate,
  cmpTime
};
