// @omnistudio/components/src/modules/c/salesforceUtils/wizardLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const closeLabel = "Close";
const cancelLabel = "Cancel";
const backLabel = "Back";
const stepLabel = "Step";
const nextLabel = "Next";
const saveLabel = "Save";
const errorToastTitle = "An error occured";
const wizItemLabelMissingError =
  "Wizard - Label is required attribute for WizardItems";

export {
  closeLabel,
  cancelLabel,
  backLabel,
  stepLabel,
  nextLabel,
  saveLabel,
  errorToastTitle,
  wizItemLabelMissingError
};
