// @omnistudio/cards/src/modules/c/flexCardUtils/dataSourceList.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const dsWizTitle = "Configure Data Source";
const dsWizSelectSourceTitle = "Select Source Type";
const dsWizSelectSourceDesc =
  "From the following options, indicate which type of datasource you would like to use.";
const dsWizSoqlName = "SOQL Query";
const dsWizSoqlDesc =
  "Use the Salesforce Object Query Language (SOQL) to search your organization’s Salesforce data for specific information.";
const dsWizSoslName = "SOSL Search";
const dsWizSoslDesc =
  "Use the Salesforce Object Search Language (SOSL) to construct text-based search queries against the search index.";
const dsWizApexRemName = "Apex Remote";
const dsWizApexRemDesc =
  "Apex Remote data source is used to make a Apex Remote call.";
const dsWizApexRestName = "Apex REST";
const dsWizApexRestDesc =
  "Apex REST data source is used to make a Apex REST call.";
const dsWizDRName = "DataRaptor";
const dsWizDRDesc =
  "DataRaptor data source is used to make calls to Vlocity DataRaptors.";
const dsWizRestName = "REST";
const dsWizRestDesc = "Uses the standard REST API call.";
const dsWizIPName = "Integration Procedures";
const dsWizIPDesc =
  "Integration Procedure data source is used to make calls to Vlocity Integration Procedures.";
const dsWizStrApiName = "Streaming API<";
const dsWizStrApiDesc =
  "Streaming API enables streaming of events using push technology and provides a subscription mechanism for receiving events in near real time.";
const dsWizCustomName = "Custom";
const dsWizCustomDesc = "Uses the sample JSON provided.";
const dsWizNoneName = "None";

export const dsObj = {
  headerTitle: dsWizTitle || "Configure Data Source",
  title: dsWizSelectSourceTitle || "Select Source Type",
  description:
    dsWizSelectSourceDesc ||
    "From the following options, indicate which type of datasource you would like to use.",
  dsList: [
    {
      name: dsWizNoneName || "None",
      label: dsWizNoneName || "None",
      description: "",
      value: "None",
      isChecked: true,
      key: "ds1"
    },
    {
      name: dsWizSoqlName || "SOQL Query",
      label: dsWizSoqlName || "SOQL Query",
      description:
        dsWizSoqlDesc ||
        "Use the Salesforce Object Query Language (SOQL) to search your organization’s Salesforce data for specific information.",
      value: "Query",
      isChecked: false,
      key: "ds2"
    },
    {
      name: dsWizSoslName || "SOSL Search",
      label: dsWizSoslName || "SOSL Search",
      description:
        dsWizSoslDesc ||
        "Use the Salesforce Object Search Language (SOSL) to construct text-based search queries against the search index.",
      value: "Search",
      isChecked: false,
      key: "ds3"
    },
    {
      name: dsWizApexRemName || "Apex Remote",
      label: dsWizApexRemName || "Apex Remote",
      description:
        dsWizApexRemDesc ||
        "Apex Remote data source is used to make a Apex Remote call.",
      value: "ApexRemote",
      isChecked: false,
      key: "ds4"
    },
    {
      name: dsWizApexRestName || "Apex REST",
      label: dsWizApexRestName || "Apex REST",
      description:
        dsWizApexRestDesc ||
        "Apex REST data source is used to make a Apex REST call.",
      value: "ApexRest",
      isChecked: false,
      key: "ds5"
    },
    {
      name: dsWizDRName || "DataRaptor",
      label: dsWizDRName || "DataRaptor",
      description:
        dsWizDRDesc ||
        "DataRaptor data source is used to make calls to Vlocity DataRaptors.",
      value: "DataRaptor",
      isChecked: false,
      key: "ds7"
    },
    {
      name: dsWizRestName || "REST",
      label: dsWizRestName || "REST",
      description: dsWizRestDesc || "Uses the standard REST API call.",
      value: "REST",
      isChecked: false,
      key: "ds8"
    },
    {
      name: dsWizIPName || "Integration Procedures",
      label: dsWizIPName || "Integration Procedures",
      description:
        dsWizIPDesc ||
        "Integration Procedure data source is used to make calls to Vlocity Integration Procedures.",
      value: "IntegrationProcedures",
      isChecked: false,
      key: "ds9"
    },
    {
      name: dsWizStrApiName || "Streaming API",
      label: dsWizStrApiName || "Streaming API",
      description:
        dsWizStrApiDesc ||
        "Streaming API enables streaming of events using push technology and provides a subscription mechanism for receiving events in near real time.",
      value: "StreamingAPI",
      isChecked: false,
      key: "ds10"
    },
    {
      name: dsWizCustomName || "Custom",
      label: dsWizCustomName || "Custom",
      description: dsWizCustomDesc || "Uses the sample JSON provided.",
      value: "Custom",
      isChecked: false,
      key: "ds11"
    }
  ]
};
