// @omnistudio/oui/src/modules/c/omniscriptTime/omniscriptTime.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
import OmniscriptAtomicElement from 'c/omniscriptAtomicElement';
import { timeFormatCompatability, parseChronoAttributes } from 'c/omniscriptUtils';
import tmpl from './omniscriptTime.html';

/**
 * Omniscript Time Element
 */
export default class OmniscriptTime extends OmniscriptAtomicElement {
    _setValueStrict = false;

    handleBlur() {}

    handleChange(evt) {
        this.applyCallResp(evt.target.value);
    }

    get defaultValue() {
        let dfltValue = super.defaultValue;
        return parseChronoAttributes(dfltValue);
    }

    validateData(data) {
        var valid = data == null || data === '';
        var asDate, useAsDate;
        if (!valid) {
            if (this.jsonDef && this._propSetMap && this._propSetMap.timeType && this._propSetMap.timeType.toUpperCase() === 'STRING') {
                valid = typeof data === 'string';
            }
            if (!valid) {
                valid = data.constructor === Date || ((asDate = new Date(data)), (useAsDate = asDate.toString() !== 'Invalid Date'), useAsDate);
            }
            if (!valid) {
                asDate = parseChronoAttributes(data);
                if (asDate != null) {
                    useAsDate = true;
                    valid = true;
                }
            }
        }
        return { valid: valid, dataToApply: useAsDate ? asDate : data };
    }

    initCompVariables() {
        super.initCompVariables();
        this._messageWhenBadInput = this.allCustomLabelsUtil.OmniValidationBadInputDate;
        if (this.jsonDef && this._propSetMap) {
            if (!this._propSetMap.strictDateFormatting) {
                if (this._propSetMap.timeFormat) {
                    this._timeFormat = timeFormatCompatability(this._propSetMap.timeFormat);
                }
                if (this._propSetMap.modelTimeFormat) {
                    this._modelTimeFormat = timeFormatCompatability(this._propSetMap.modelTimeFormat);
                }
            }
            if (this._propSetMap.minTime) {
                this._min = parseChronoAttributes(this._propSetMap.minTime);
            }
            if (this._propSetMap.maxTime) {
                this._max = parseChronoAttributes(this._propSetMap.maxTime);
            }

            this._timeInterval = this._propSetMap.timeInterval || 30;
            this._messageWhenRangeOverflow = String(this.allCustomLabelsUtil.OmniTimeMinMax) + ' %startTime% - %endTime%';
            this._messageWhenRangeUnderflow = this._messageWhenRangeOverflow;
        }
    }

    setElementFormattedValue() {
        if (this.childInput) {
            const dispVal = this.childInput.displayValue;
            // odd structure to only call getter once;
            if (dispVal != null) {
                this._elementFormattedValue = dispVal;
            }
        }
        super.setElementFormattedValue();
    }

    render() {
        return tmpl;
    }
}
