// @omnistudio/oui/src/modules/c/omniscriptCurrency/omniscriptCurrency.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
/**
 * @module ns/omniscriptCurrency
 * @description This component is used to render a currency Element, This is extends from class `OmniscriptAtomicElement`.
 */
import OmniscriptAtomicElement from 'c/omniscriptAtomicElement';
import tmpl from './omniscriptCurrency.html';

/**
 * Default exported class OmniscriptCurrency.
 * @extends OmniscriptAtomicElement
 * @typicalname OmniscriptCurrency
 */
export default class OmniscriptCurrency extends OmniscriptAtomicElement {
    /**
     * @type {Object} - Hold an object from Helper method for mask handling
     * @scope private
     */
    _imaskCurrencyAttributes;

    /**
     * @type {Boolean} - Checks whether value is committed or not
     * @scope private
     */
    _commitOnChange;

    /**
     * @description Overwrites inherited initCompVariables.
     * @returns {Void}
     * @scope private
     */
    initCompVariables() {
        super.initCompVariables();
        this._imaskCurrencyAttributes = this.getImaskCurrencyAttributes();
        this._commitOnChange = this.scriptHeaderDef.propSetMap.commitOnChange === true;
        this._repeatCurrencyAlternativeText = this.allCustomLabelsUtil.OmniCurrencyRepeatAltText.replace(/\{0\}/gi, this._propSetMap.label);
        this._removeCurrencyAlternativeText = this.allCustomLabelsUtil.OmniCurrencyRemoveAltText.replace(/\{0\}/gi, this._propSetMap.label);
    }

    /**
     * Event handler function on blur event.
     * Sets the element value and triggers aggregation
     * @returns {void}
     * @scope private
     */
    handleBlur(evt) {
        this.applyCallResp(evt.target.value);
    }

    /**
     * @description Evaluates if currency is valid.
     * @param {Number} data
     * @returns {Object}
     * @scope private
     */
    validateData(data) {
        // default - Number
        let valid = typeof data === 'number' || data === null ? true : false;
        if (valid && data < 0 && !this._propSetMap.allowNegative) {
            valid = false;
            data = null;
        }
        return { valid: valid, dataToApply: data };
    }

    /**
     * @scope private
     * @description Overwrites the native LWC render.
     * @returns {Template}
     */
    render() {
        return tmpl;
    }
}
